tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1166 @@
+    /// }
+    ///
+    /// // A functional, unidirectional server-to-client only communication.
+    /// {
+    ///     let mut server = ServerOptions::new()
+    ///         .access_inbound(false)
+    ///         .create(PIPE_NAME)?;
+    ///
+    ///     let mut client = ClientOptions::new()
+    ///         .write(false)
+    ///         .open(PIPE_NAME)?;
+    ///
+    ///     let write = server.write_all(b"ping");
+    ///
+    ///     let mut buf = [0u8; 4];
+    ///     let read = client.read_exact(&mut buf);
+    ///
+    ///     let ((), read) = tokio::try_join!(write, read)?;
+    ///
+    ///     assert_eq!(read, 4);
+    ///     assert_eq!(&buf[..], b"ping");
+    /// }
+    /// # Ok(()) }
+    /// ```
+    pub fn access_inbound(&mut self, allowed: bool) -> &mut Self {
+        bool_flag!(self.open_mode, allowed, winbase::PIPE_ACCESS_INBOUND);
+        self
+    }
+
+    /// The flow of data in the pipe goes from server to client only.
+    ///
+    /// This corresponds to setting [`PIPE_ACCESS_OUTBOUND`].
+    ///
+    /// [`PIPE_ACCESS_OUTBOUND`]: https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createnamedpipea#pipe_access_outbound
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::io;
+    /// use tokio::io::{AsyncReadExt as _, AsyncWriteExt as _};
+    /// use tokio::net::windows::named_pipe::{ClientOptions, ServerOptions};
+    ///
+    /// const PIPE_NAME: &str = r"\\.\pipe\tokio-named-pipe-access-outbound";
+    ///
+    /// # #[tokio::main] async fn main() -> io::Result<()> {
+    /// // Server side prevents connecting by denying outbound access, client errors
+    /// // when attempting to create the connection.
+    /// {
+    ///     let _server = ServerOptions::new()
+    ///         .access_outbound(false)
+    ///         .create(PIPE_NAME)?;
+    ///
+    ///     let e = ClientOptions::new()
+    ///         .open(PIPE_NAME)
+    ///         .unwrap_err();
+    ///
+    ///     assert_eq!(e.kind(), io::ErrorKind::PermissionDenied);
+    ///
+    ///     // Disabling reading allows a client to connect, but leads to runtime
+    ///     // error if a read is attempted.
+    ///     let mut client = ClientOptions::new()
+    ///         .read(false)
+    ///         .open(PIPE_NAME)?;
+    ///
+    ///     let mut buf = [0u8; 4];
+    ///     let e = client.read(&mut buf).await.unwrap_err();
+    ///     assert_eq!(e.kind(), io::ErrorKind::PermissionDenied);
+    /// }
+    ///
+    /// // A functional, unidirectional client-to-server only communication.
+    /// {
+    ///     let mut server = ServerOptions::new().access_outbound(false).create(PIPE_NAME)?;
+    ///     let mut client = ClientOptions::new().read(false).open(PIPE_NAME)?;
+    ///
+    ///     // TODO: Explain why this test doesn't work without calling connect
+    ///     // first.
+    ///     //
+    ///     // Because I have no idea -- udoprog
+    ///     server.connect().await?;
```

Comment:
Is it not just because you need to call `connect()` before you can use any connections?

Reply:
If clients are (successfully) created after servers they start out being connected. `connect` is just supposed to be equivalent to waiting for a client to connect.

I've tried this sequence outside of Tokio and Mio and it works there (It was a while ago now, so I'm gonna double check this). I'm gonna try to hunt down what is actually going on because it is a bit concerning.

For now I'm removing this comment and ensure that we call `connect`.

I'll also be breaking up the doctest in the process. It's a bit messy to have it all in one instead of separating errors and examples.